core-primitives/stf-state-handler/src/state_snapshot_repository.rs

```diff
@@ -0,0 +1,480 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+use crate::{
+	error::{Error, Result},
+	file_io::StateFileIo,
+};
+use core::ops::RangeBounds;
+use itp_settings::files::ENCRYPTED_STATE_FILE;
+use itp_time_utils::now_as_nanos;
+use itp_types::ShardIdentifier;
+use log::*;
+use std::{
+	collections::{HashMap, VecDeque},
+	fmt::Debug,
+	format,
+	marker::PhantomData,
+	string::{String, ToString},
+	sync::Arc,
+	vec::Vec,
+};
+
+/// Trait for versioned state access. Manages history of state snapshots.
+pub trait VersionedStateAccess {
```

Comment:
This is the trait with the new functionality to store a history of state snapshots. It is still quite basic for now, since we're not really using the `revert` mechanism yet for rollbacks.